IHP/Log/Types.hs

```diff
@@ -0,0 +1,154 @@
+{-|
+Module: IHP.Log.Types
+Description:  Types for the IHP logging system
+-}
+{-# LANGUAGE GADTs #-}
+
+module IHP.Log.Types (
+  module System.Log.FastLogger,
+  Logger(..),
+  LogLevel(..),
+  LogDestination(..),
+  LoggingProvider(..),
+  LoggerSettings(..),
+  LogFormatter,
+  FormattedTime,
+  newLogger,
+  defaultLogger,
+  defaultDestination,
+  defaultFormatter,
+  withLevelFormatter,
+  withTimeFormatter,
+  withTimeAndLevelFormatter,
+) where
+
+import IHP.HaskellSupport
+import qualified Prelude
+import CorePrelude hiding (putStr, putStrLn, print, error, show)
+import Data.Text as Text
+import Data.Default.Class (Default (def))
+import System.Log.FastLogger (
+  LogStr,
+  LogType'(..),
+  BufSize,
+  FileLogSpec(..),
+  TimedFileLogSpec(..),
+  TimeFormat,
+  newFastLogger,
+  toLogStr,
+  fromLogStr,
+  defaultBufSize,
+  newTimeCache,
+  simpleTimeFormat'
+  )
+
+import qualified System.Log.FastLogger as FastLogger (FormattedTime)
+
+
+-- some functions brought over from IHP.Prelude
+-- can't import due to circular dependency with IHP.ModelSupport which relies on this module
+
+tshow :: Show a => a -> Text
+tshow value = Text.pack (Prelude.show value)
+
+show :: Show a => a -> Text
+show = tshow
+
+data Logger = Logger {
+  write     :: Text -> IO (),
+  level     :: LogLevel,
+  formatter :: LogFormatter,
+  timeCache :: IO FastLogger.FormattedTime,
+  cleanup   :: IO ()
+}
+
+data LogLevel =
+  Debug
+  | Info
+  | Warn
+  | Error
+  | Fatal
+  | Unknown
+  deriving (Enum, Eq, Ord, Show)
+
+type FormattedTime = Text
+type LogFormatter = FormattedTime -> LogLevel -> Text -> Text
+
+-- | Where logged messages will be delivered to. Types correspond with those in fast-logger.
+data LogDestination where
```

Comment:
Why are we using GADTs here?

Reply:
I was also just wondering: Why aren't we just reexporting fast loggers data types here?

Reply:
GADTs are unnecessary -- switched back to regular data type. Thought it might be necessary for some reason when I started but didn't turn out that way!

I chose to wrap the fast logger types for a couple reasons:
- allow for using alternate logging back-ends in the future by decoupling the user facing IHP code from the dependency
- allow for IHP to document the logger types in its own documentation 
- allow for wrapped configuration types, for example the FileLogSpec and TimedFileLogSpec records are a little unwieldy (and use snake case for the field names???)

That being said the code I wrote here doesn't take much advantage of these benefits: although the types are wrapped they still use fast-logger types to configure them so they aren't decoupled in any meaningful way, and the documentation I have isn't any better than the fast-logger docs.

I'd be happy to simply use the fast-logger types if you'd prefer, or I can also decouple in a more meaningful way and work on extra documentation to get more of a benefit from a custom IHP type. Let me know :)